File: .config/waybar/spotify.ts
#!/usr/bin/env bun

/**
 * Spotify Waybar Module
 *
 * Usage:
 *   bun spotify.ts                    # Get current status (for display)
 *   bun spotify.ts --action toggle    # Play/Pause
 *   bun spotify.ts --action next      # Next track
 *   bun spotify.ts --action previous  # Previous track
 */

const PLAYER = 'spotify';

interface WaybarOutput {
  text: string;
  tooltip?: string;
  class?: string;
  alt?: string;
}

async function runCommand(cmd: string[]): Promise<string> {
  const proc = Bun.spawn(cmd, {
    stdout: 'pipe',
    stderr: 'pipe',
  });

  const output = await new Response(proc.stdout).text();
  const error = await new Response(proc.stderr).text();

  if (proc.exitCode !== 0 && error) {
    throw new Error(error.trim());
  }

  return output.trim();
}

async function isPlayerRunning(): Promise<boolean> {
  try {
    const players = await runCommand(['playerctl', '-l']);
    return players.includes(PLAYER);
  } catch {
    return false;
  }
}

async function getMetadata(): Promise<{
  title: string;
  artist: string;
  album: string;
  status: 'Playing' | 'Paused' | 'Stopped';
}> {
  const [title, artist, album, status] = await Promise.all([
    runCommand(['playerctl', '-p', PLAYER, 'metadata', 'title']).catch(
      () => '',
    ),
    runCommand(['playerctl', '-p', PLAYER, 'metadata', 'artist']).catch(
      () => '',
    ),
    runCommand(['playerctl', '-p', PLAYER, 'metadata', 'album']).catch(
      () => '',
    ),
    runCommand(['playerctl', '-p', PLAYER, 'status']).catch(() => 'Stopped'),
  ]);

  return {
    title,
    artist,
    album,
    status: status as 'Playing' | 'Paused' | 'Stopped',
  };
}

async function executeAction(action: string): Promise<void> {
  if (!(await isPlayerRunning())) {
    return;
  }

  switch (action) {
    case 'toggle':
      await runCommand(['playerctl', '-p', PLAYER, 'play-pause']);
      break;
    case 'next':
      await runCommand(['playerctl', '-p', PLAYER, 'next']);
      break;
    case 'previous':
      await runCommand(['playerctl', '-p', PLAYER, 'previous']);
      break;
  }
}

function truncateText(text: string, maxLength: number): string {
  if (text.length <= maxLength) return text;
  return text.slice(0, maxLength - 3) + '...';
}

async function getStatusOutput(): Promise<WaybarOutput> {
  if (!(await isPlayerRunning())) {
    return { text: '' };
  }

  const { title, artist, album, status } = await getMetadata();

  if (!title) {
    return { text: '' };
  }

  const isPlaying = status === 'Playing';
  const statusIndicator = isPlaying ? '⏸' : '▶';
  const displayText = truncateText(
    `${statusIndicator} ${title} - ${artist}`,
    35,
  );
  const tooltip = `Música: ${title}\nArtista: ${artist}${album ? `\nÁlbum: ${album}` : ''}\nStatus: ${status}`;

  return {
    text: displayText,
    tooltip,
    class: isPlaying ? 'playing' : 'paused',
    alt: status.toLowerCase(),
  };
}

async function main() {
  const args = process.argv.slice(2);
  const actionIndex = args.indexOf('--action');

  if (actionIndex !== -1 && args[actionIndex + 1]) {
    await executeAction(args[actionIndex + 1]);
    return;
  }

  const output = await getStatusOutput();
  console.log(JSON.stringify(output));
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: .config/waybar/spotify.lua
#!/usr/bin/env lua

-- Usage:
--   lua spotify.lua                    # Get current status
--   lua spotify.lua --action toggle    # Play/Pause
--   lua spotify.lua --action next      # Next track
--   lua spotify.lua --action previous  # Previous track

local PLAYER = "spotify"
local MAX_TEXT_LENGTH = 45

local function run_command(cmd)
    local handle = io.popen(cmd .. " 2>/dev/null")
    if not handle then
        return ""
    end
    local result = handle:read("*a")
    handle:close()
    return result:gsub("^%s+", ""):gsub("%s+$", "")
end

local function is_running()
    local players = run_command("playerctl -l")
    return players:find(PLAYER) ~= nil
end

local function get_metadata()
    local title = run_command("playerctl -p " .. PLAYER .. " metadata title")
    local artist = run_command("playerctl -p " .. PLAYER .. " metadata artist")
    local album = run_command("playerctl -p " .. PLAYER .. " metadata album")
    local status = run_command("playerctl -p " .. PLAYER .. " status")

    if status ~= "Playing" and status ~= "Paused" then
        status = "Stopped"
    end

    return { title = title, artist = artist, album = album, status = status }
end

local function execute_action(action)
    if not is_running() then return end

    if action == "toggle" then
        run_command("playerctl -p " .. PLAYER .. " play-pause")
    elseif action == "next" then
        run_command("playerctl -p " .. PLAYER .. " next")
    elseif action == "previous" then
        run_command("playerctl -p " .. PLAYER .. " previous")
    end
end

-- Escape special chars for JSON/Pango
local function escape_text(text)
    text = text:gsub('"', '\\"')
    text = text:gsub("<", "&lt;")
    text = text:gsub(">", "&gt;")
    return text
end

local function truncate(text, max_length)
    if #text <= max_length then
        return text
    end
    return text:sub(1, max_length - 3) .. "..."
end

local function get_output()
    if not is_running() then
        print('{"text": ""}')
        return
    end

    local meta = get_metadata()

    if meta.title == "" then
        print('{"text": ""}')
        return
    end

    local is_playing = meta.status == "Playing"
    local indicator = is_playing and "⏸" or "▶"
    local display = truncate(indicator .. " " .. meta.title .. " - " .. meta.artist, MAX_TEXT_LENGTH)

    display = escape_text(display)
    meta.title = escape_text(meta.title)
    meta.artist = escape_text(meta.artist)
    meta.album = escape_text(meta.album)

    local class = is_playing and "playing" or "paused"
    local alt = meta.status:lower()

    local tooltip = "Music: " .. meta.title ..
        "\\nArtist: " .. meta.artist ..
        (meta.album ~= "" and ("\\nAlbum: " .. meta.album) or "") ..
        "\\nStatus: " .. meta.status

    tooltip = escape_text(tooltip)

    local json = string.format(
        '{"text": "%s", "tooltip": "%s", "class": "%s", "alt": "%s"}',
        display, tooltip, class, alt
    )

    print(json)
end

local args = arg or {}
local action_index = nil

for i = 1, #args do
    if args[i] == "--action" then
        action_index = i
        break
    end
end

if action_index and args[action_index + 1] then
    execute_action(args[action_index + 1])
else
    get_output()
end
